src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -992,8 +992,10 @@ protected Task TryProduceInvalidRequestResponse()
         {
             return ProduceEnd();
         }
-
-        return Task.CompletedTask;
+        else
+        {
+            return Output.FlushAsync(CancellationToken.None).GetAsTask();
```

Comment:
I'm not sure about this. Need somewhere to flush the HTTP/2 GOAWAY frame content written in TryParseRequest.

Reply:
It does feal weird to flush a response when `_connectionAborted` is true. If we stick with writing a GOAWAY for h2c connections to an HTTP/1.1 endpoint, I'd probably just make this virtual (it only gets called once per connection anyway) and use a new flag in `DetectHttp2Preface` (`_isInvalidH2cConnection`?). If set, I'd write the `Http2GoAwayProtocolErrorBytes` here and flush it and skip the default behavior. Otherwise, I'd just call into the base implementation.

That way you wouldn't have to set `_connectionAborted` in `DetectHttp2Preface` either which feels like a win.